### DISCUSSION TITLE
Trailing semi-colon

### DISCUSSION THREAD

**Discussion by prescod** [kdl-org/kdl#187] [Ideas]
Hello! KDL looks interesting!

One of the benefits of having closing curly bases should be that it is obvious when a statement ends. 

Unfortunately the KDL grammar seems to force me to explicitly end the statement (at least in my experiments and my skim of the grammar:

object1 {object2;}

Rather than:

object1 {object2}

I had thought that the use of semi-colon as terminator, rather than separator, had gone out of fashion. At least in languages where semicolons are usually optional.

What benefit does the end-user get from the grammar requiring a statement terminator? Would you consider making it optional, even at a cost of complicating the text of the grammar?

**Comment by hkolbeck** [comment]
That's only the case when placing the terminating brace on the same line as a child node. I think it could be worth considering to allow the latter example, but encouraging placement of children on the same line feels like it could be an anti-pattern. At the end of the day, requiring a single character extra in a corner case doesn't feel like it's terribly user-impacting.

**Comment by jihchi** [comment]
Just FYI. I found there is an issue discussing the same thing: https://github.com/kdl-org/kdl-rs/issues/11 .

The issue seems like a spec compliance bug if I understand the issue above correctly.

**Reply by hkolbeck** [reply]
I'm a little confused, I don't see how `node-terminator := single-line-comment | newline | ';' | eof` allows a child on the same line to not have an explicit `;` terminator?

**Reply by hkolbeck** [reply]
kdl4j will also reject `node {child}` currently, I'll add a test to the suite whichever direction is decided on.

**Comment by zkat** [comment]
What's the proposed grammar for this?

**Reply by hkolbeck** [reply]
I think it'd need to be something like:

```
unterminated-node := ('/-' node-space*)? type? identifier (node-space node-space* node-props-and-args)* (node-space* node-children ws*)? node-space*
node := unterminated-node node-terminator
node-children := ('/-' node-space*)?  '{' nodes  unterminated-node? '}'
```

I'm not confident in this. If this is what it takes, I stand by my earlier feeling that it's not worthwhile to save a single character in an edge case.

**Reply by jihchi** [reply]
Current grammar looks fine to me, I thought it is a bug in the implementation. 

Please allow me to paste (partial) grammar here (see [here](https://github.com/kdl-org/kdl/blob/main/SPEC.md#full-grammar) for full grammar):

```bnf
nodes := linespace* (node nodes?)? linespace*

node := ('/-' node-space*)? type? identifier (node-space node-space* node-props-and-args)* (node-space* node-children ws*)? node-space* node-terminator
node-props-and-args := ('/-' node-space*)? (prop | value)
node-children := ('/-' node-space*)? '{' nodes '}'
node-space := ws* escline ws* | ws+
node-terminator := single-line-comment | newline | ';' | eof
```

Suppose we have a KDL document with one top-level node:

```kdl
object1 {object2}
```

Its parser tree looks like this:

```
              ┌──────┐                                 
              │ node │                                 
              └──────┘                                 
                  │                                    
                  ▼                                    
                                                       
         "object1 {object2}"                           
          │     ││        │                            
          └──┬──┘└────┬───┘                            
       ┌─────┘        └───────────┐                    
       ▼                          ▼                    
┌────────────┐  ┌──────────────────────────────────┐   
│ identifier │  │ (node-space* node-children ws*)? │   
└────────────┘  └──────────────────────────────────┘   
       │                          │                    
       ▼                          ▼                    
                                                       
   "object1"                " {object2}"               
                             ││       │                
                      ┌──────┘└───┬───┘                
                      │           └─────┐              
                      ▼                 ▼              
               ┌─────────────┐  ┌───────────────┐      
               │ node-space* │  │ node-children │      
               └─────────────┘  └───────────────┘      
                      │                 │              
                      ▼                 ▼              
                                                       
                     " "           "{object2}"         
                                    ││     ││          
                             ┌──────┘└──┬──┘└──────┐   
                             │          │          │   
                             ▼          ▼          ▼   
                          ┌─────┐   ┌───────┐   ┌─────┐
                          │ '{' │   │ nodes │   │ '}' │
                          └─────┘   └───────┘   └─────┘
                                        │              
                                        ▼              
                                 ┌────────────┐        
                                 │ identifier │        
                                 └────────────┘        
                                        │              
                                        ▼              
                                                       
                                    "object2"          
                                                       
```

Am I interpreting it correctly?

**Reply by hkolbeck** [reply]
I think the issue is that that diagram is missing several steps between `nodes` and `identifier` at the bottom. Importantly, one of them includes the non-optional token `node-terminator`.

**Reply by jihchi** [reply]
I've simplified some (minor) steps for the sake of simplicity.

Isn't that "empty" input means `eof` in `node-terminator`?

**Reply by hkolbeck** [reply]
`eof` only matches the end of the entire file, not the block.

**Reply by jihchi** [reply]
How about adding an empty string `''` to `node-terminator`?

```
node-terminator := single-line-comment | newline | ';' | eof | ''
```

**Reply by hkolbeck** [reply]
Adding an empty string to node-terminator is basically the same as changing `node` to end with `node-terminator?`. Doing that wouldn't work, because imagine a document like:

```
node1 "arg1" "arg2"
```

If terminator is optional, you could parse this as a node with two args, three nodes with no args, etc, and all would match the grammar. 

The only approach I see what what I sketched up above, which basically says that the last node in a child block can be unterminated. That's doable because the `}` token can't be matched against any of the sub-parts of `node`. I'm not an expert, and I definitely don't promise that there aren't simpler approaches.

**Reply by larsgw** [reply]
> ```
> '{' nodes unterminated-node? '}'
> ```

Maybe whitespace at the end.

```
'{' nodes (unterminated-node ws*)? '}'
```

**Reply by hkolbeck** [reply]
After some additional thought, the definition I gave is a terrible idea, because it makes it ambiguous which of nodes and unterminated-node? would match the lone child node with a semicolon. This seems like it could cause issues creating parsers for KDL in many parser generators. If I'm mistaken in that, then I withdraw my withdrawl.

**Reply by larsgw** [reply]
No, that's probably true. I can look into combining EOF and "no terminator" later today.

**Reply by hkolbeck** [reply]
I want to make it explicit again that I think this change to the language is unnecessary, and unless it can be done pretty simply I'm not in favor of making it.

**Reply by zkat** [reply]
I agree. I just don't think we should do it. It's not worth the breaking change, and it's not worth the increased implementation complexity and lookahead requirement.
